### PR TITLE
Improve `API` class, add retry with `tenacity` (#4)

### LINKED ISSUE #4: Handle occasional invalid JSON or timeouts from Canvas using retries
Given the number of API calls we're making during updating, we likely need to handle the occasional irregular status code. In the past, we've seen gateway timeouts and some incomplete JSON responses from Canvas. 

We've written custom code to handles this before in Placement Exams and IPE, but something like [tenacity](https://github.com/jd/tenacity) could do the trick.

### PR DESCRIPTION
**Pull request by ssciolla**
This PR aims to resolve #4.

To just test `API`:
```
python3 migration/tests.py APITestCase -v
```

To Do
- [x]  Improve `limit` logic
- [x] Create `API.get` and `API.put`.
- [x] Apply `tenacity.retry` to each
- [x] Add test case for `API`
- [x] Move some tests to the proper test case

### CODE REVIEW

File: migration/api.py [hunk 1]
```diff
@@ -14,22 +20,58 @@ class EndpointType(Enum):
...
+        before_sleep=before_sleep_log(logger, logging.WARN)
+    )
+    def put(self, url: str, params: dict[str, Any] | None = None) -> Any:
+        resp = self.client.put(url=url, params=params)
+        resp.raise_for_status()
+        return resp.json()
+
     def get_results_from_pages(
-        self, endpoint: str, params: dict[str, Any] | None = None, page_size: int = 50, bail_after: int | None = None
+        self, endpoint: str, params: dict[str, Any] | None = None, page_size: int = 50, limit: int | None = None
```
Comment (ssciolla) [inline migration/api.py:74]: I decided to rename `bail_after` to `limit`, seems more straightforward and intuitive.

File: migration/api.py [hunk 2]
```diff
@@ -44,22 +86,19 @@ def get_results_from_pages(
...
-            elif bail_after is not None and bail_after <= len(results):
+            elif limit is not None and limit <= len(results):
                 more_pages = False
             else:
-                extra_params.update(next_page_params)
+                extra_params.update(data.next_page_params)
                 page_num += 1
 
+        if limit is not None and len(results) > limit:
+            results = results[:limit]
```
Comment (ssciolla) [inline migration/api.py:100]: If `limit` is specified, I'm now actually slicing off any extra data, which I wasn't doing before.

File: migration/manager.py [hunk 3]
```diff
@@ -22,18 +21,19 @@ def get_tools_installed_in_account(self) -> list[ExternalTool]:
...
         return tools
 
-    def get_courses_in_terms(self, term_ids: list[int], bail_after: int | None = None) -> list[Course]:
-        limit_per_term = None
-        if bail_after is not None:
-            limit_per_term = bail_after // len(term_ids)
+    def get_courses_in_terms(self, term_ids: list[int], limit: int | None = None) -> list[Course]:
+        limit_chunks = None
+        if limit is not None:
+            limit_chunks = chunk_integer(limit, len(term_ids))
```
Comment (ssciolla) [inline migration/manager.py:27]: I added a utility function and logic so that `limit` could be used for this method with multiple terms. `chunk_integer` ensures that -- except for some edge cases -- the result should have courses from each term. If `limit` isn't specified, there should not be any effect, all results will be returned.

File: migration/tests.py [hunk 4]
```diff
@@ -1,49 +1,124 @@
...
+        bad_resp = httpx.Response(
+            status_code=httpx.codes.BAD_GATEWAY,
+            request=request
+        )
+        with patch.object(self.api.client, 'put', autospec=True) as mock_put_call:
+            mock_put_call.side_effect = [bad_resp, bad_resp, bad_resp, bad_resp]
+            with self.api.client:
+                with self.assertRaises(httpx.HTTPStatusError):
+                    self.api.put(self.course_url, params={ "name": "Test Course!" })
+        self.assertEqual(mock_put_call.call_count, 4)
```
Comment (ssciolla) [inline migration/tests.py:104]: I could definitely add more tests here, but I'm going to leave it like this for now. I think the retrying bit is pretty covered, but success cases for `get` and `put` could be added.

File: migration/tests.py [hunk 5]
```diff
@@ -93,6 +177,22 @@ def setUp(self):
...
+            position=30
+        )
+
+    def test_find_tab_by_tool_id_returns_tab(self):
+        tab = CourseManager.find_tab_by_tool_id(99999, [self.test_external_tool_tab])
+        self.assertTrue(isinstance(tab, ExternalToolTab))
+
+    def test_find_tab_by_tool_id_returns_none(self):
+        tab = CourseManager.find_tab_by_tool_id(100000, [self.test_external_tool_tab])
+        self.assertTrue(tab is None)
```
Comment (ssciolla) [inline migration/tests.py:194]: I'm just moving these from `AccountManagerTestCase`, they were in the wrong place.

File: migration/api.py [hunk 6]
```diff
@@ -14,22 +20,58 @@ class EndpointType(Enum):
...
+    )
+    def get(self, url: str, params: dict[str, Any] | None = None) -> GetResponseData:
+        resp = self.client.get(url=url, params=params)
+        resp.raise_for_status()
+        data = resp.json()
+        next_page_params = self.get_next_page_params(resp)
+        return GetResponseData(data, next_page_params)
+
+    @retry(
+        stop=stop_after_attempt(4),
```
Comment (ssciolla) [inline migration/api.py:63]: I think it's probably possible to make this configurable, but having done so in the past with other projects, this doesn't seem like something we ever change. If we did, I think it'd just have to be a environment variable/global variable in this file, since there isn't a good way to pass class instance variables to decorators.

Comment (zqian) [reply migration/api.py:63]: We can leave the retry number hard coded for now. You can remove this item from the checklist.

File: migration/api.py (hunk 2 above)
Comment (ssciolla) [reply migration/api.py:101]: Sure, done. https://github.com/tl-its-umich-edu/tool-migration/pull/7/commits/48c65e27d544ef050b40d16ebe5925cbde2c77ac

**Comment by ssciolla** [comment]
Hi @zqian, were you still reviewing this?

File: migration/api.py (hunk 2 above)
Comment (zqian) [reply migration/api.py:101]: Thanks!

File: migration/tests.py (hunk 4 above)
Comment (zqian) [reply migration/tests.py:104]: Sounds good.

**Review by zqian — APPROVED** [review]
